azurerm_monitor_diagnostic_setting - added LogAnalyticsDestinationType

KimShepherdVibrato opened the pull request:
Added the argument log_analytics_destination_type to the resource azurerm_monitor_diagnostic_setting.

This allows logs sent to a Log Analytics Workspace to go to dedicated resource tables, instead of to the AzureDiagnostics table.  This is only relevant to Azure Data Factory right now, but is planned to be extended to all resources over time.

https://azure.microsoft.com/en-au/updates/adf-logs-in-dedicated-tables/
https://docs.microsoft.com/en-us/azure/azure-monitor/platform/diagnostic-logs-stream-log-store#azure-diagnostics-vs-resource-specific

katbyte commented on azurerm/resource_arm_monitor_diagnostic_setting.go:
```diff
@@ -73,6 +73,19 @@ func resourceArmMonitorDiagnosticSetting() *schema.Resource {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: false,
+				ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
```
I believe this could be:
```suggestion
				ValidateFunc: validation.StringInSlice([]string{"Dedicated"}),
```

KimShepherdVibrato replied:
I've made the change to the ValidateFunc.  The property seems to be case sensitive so I've set ignoreCase to false

katbyte commented on azurerm/resource_arm_monitor_diagnostic_setting.go:
```diff
@@ -234,6 +247,11 @@ func resourceArmMonitorDiagnosticSettingCreateUpdate(d *schema.ResourceData, met
 	}
 
+	logAnalyticsDestinationType := d.Get("log_analytics_destination_type").(string)
+	if logAnalyticsDestinationType != "" {
```
This could be simplified to
```suggestion
	if v := d.Get("log_analytics_destination_type").(string); v != "" {
		properties.DiagnosticSettings.LogAnalyticsDestinationType = &v
	}
```

katbyte replied:
Could we also check that if a type is specified that a `log_analytics_workspace_id` is too?

KimShepherdVibrato replied:
I've made the simplification and added a check that the workspaceId is set.

KimShepherdVibrato commented:
I've added the log_analytics_destination_type to the logAnalyticsWorkspace.  Do you think this is adequate, or should I create a dedicated test?

KimShepherdVibrato commented:
Hi @katbyte ,

I must admit the test changes were a bit of Cargo Cult Coding, I didn't really understand and I think I must have just run the unit test suite.

I thought I had tested using the log_analytics_destination_type with a resource it wasn't enabled for, but I must have been mistaken as trying again this morning I get an empty string when refreshing on a key vault.

I've reverted the change to the existing test and added a new test: TestAccAzureRMMonitorDiagnosticSetting_logAnalyticsWorkspaceDedicated

This test uses an Azure Datafactory to correctly test the Dedicated diagnostics.  I've confirmed the test passes for me.

tombuildsstuff commented:
Tests pass:

<img width="506" alt="Screenshot 2019-08-09 at 10 50 16" src="https://user-images.githubusercontent.com/666005/62770722-81abf580-ba93-11e9-9a08-d45ea71c5cbc.png">
